src/overlays/actors/ovl_En_Vali/z_en_vali.c

```diff
@@ -76,71 +82,732 @@ static ColliderCylinderInit D_80B288F0 = {
+    this->timer = 36;
+    this->actionFunc = EnVali_Frozen;
+}
+
+void EnVali_SetupReturnToLurk(EnVali* this) {
+    Animation_MorphToPlayOnce(&this->skelAnime, &gBariLurkingAnim, 10.0f);
+    this->actor.flags |= 0x10;
+    this->actor.flags &= ~1;
+    this->actionFunc = EnVali_ReturnToLurk;
+}
+
+void EnVali_DischargeLightning(EnVali* this, GlobalContext* globalCtx) {
+    static Color_RGBA8 effectPrimColor = { 255, 255, 255, 255 };
+    static Color_RGBA8 effectEnvColor = { 200, 255, 255, 255 };
+    Vec3f effectPos;
+    s32 i;
+    f32 cos;
+    f32 sin;
+    s16 effectYaw;
+
+    for (i = 0; i < 4; i++) {
+        cos = -Math_CosS(Camera_GetCamDirYaw(ACTIVE_CAM));
+        sin = Math_SinS(Camera_GetCamDirYaw(ACTIVE_CAM));
+        if (!((this->lightningTimer + (i << 1)) % 4)) {
+            effectYaw = (s16)(s32)Rand_CenteredFloat(12288.0f) + (i * 0x4000) + 0x2000;
+            effectPos.x = this->actor.world.pos.x + (Math_SinS(effectYaw) * 12.0f * cos);
+            effectPos.y = this->actor.world.pos.y - (Math_CosS(effectYaw) * 12.0f) + 10.0f;
+            effectPos.z = this->actor.world.pos.z + (Math_SinS(effectYaw) * 12.0f * sin);
+
+            EffectSsLightning_Spawn(globalCtx, &effectPos, &effectPrimColor, &effectEnvColor, 17, effectYaw, 6, 2);
+        }
+    }
+
+    func_8002F974(&this->actor, NA_SE_EN_BIRI_SPARK - SFX_FLAG);
+}
+
+void EnVali_Lurk(EnVali* this, GlobalContext* globalCtx) {
+    if (this->actor.xzDistToPlayer < 150.0f) {
+        EnVali_SetupDropAppear(this);
+    }
+}
+
+void EnVali_DropAppear(EnVali* this, GlobalContext* globalCtx) {
+    SkelAnime_Update(&this->skelAnime);
+    this->actor.velocity.y *= 1.5f;
+    this->actor.velocity.y = CLAMP_MAX(this->actor.velocity.y, 40.0f);
+
+    if (Math_StepToF(&this->actor.world.pos.y, this->actor.floorHeight, this->actor.velocity.y)) {
+        Audio_PlayActorSound2(&this->actor, NA_SE_EN_DODO_M_GND);
+        EnVali_SetupFloatIdle(this);
+    }
+}
+
+void EnVali_FloatIdle(EnVali* this, GlobalContext* globalCtx) {
+    s32 curFrame;
+
+    SkelAnime_Update(&this->skelAnime);
+
+    if (this->slingshotReactionTimer != 0) {
+        this->slingshotReactionTimer--;
+    }
+
+    curFrame = this->skelAnime.curFrame;
+
+    Math_StepToF(&this->floatHomeHeight, this->actor.floorHeight + 40.0f, 1.2f);
+    this->actor.world.pos.y = this->floatHomeHeight - (sinf(curFrame * M_PI * 0.0125f) * 8.0f);
+
+    if (!!this->slingshotReactionTimer) {
+        this->actor.shape.rot.y += 0x800;
+
+        if (((this->slingshotReactionTimer % 6) == 0) && (curFrame > 15) && (curFrame <= 55)) {
+            Audio_PlayActorSound2(&this->actor, NA_SE_EN_BARI_ROLL);
+        }
+    } else if ((curFrame == 16) || (curFrame == 30) || (curFrame == 42) || (curFrame == 55)) {
+        Audio_PlayActorSound2(&this->actor, NA_SE_EN_BARI_ROLL);
+    }
+
+    curFrame = ((curFrame > 40) ? (80 - curFrame) : curFrame);
+
+    this->actor.shape.rot.y += (s16)(s32)((curFrame + 4) * 0.4f * 182.04445f);
```

Comment:
Can the `s32` cast be dropped?
```suggestion
    this->actor.shape.rot.y += (s16)((curFrame + 4) * 0.4f * (0x10000 / 360.0f));
```